internal/work/dispatcher.go

```diff
@@ -328,3 +328,26 @@ func (d *Dispatcher) senderName(sender dbus.Sender) (string, error) {
 
 	return "", fmt.Errorf("cannot get name for sender: %v", sender)
 }
+
+// CancelMessage implements the dispatch of a cancel method to the worker.
+func (d *Dispatcher) CancelMessage(command yggdrasil.Command, id string) error {
```

Comment:
Following comment seems little bit off.

Reply:
This method signature feels odd to me. It strikes me as odd that we expect an
entire `yggdrasil.Command` structure as the first parameter, but then only the
message ID from the encapsulating message as the second parameter. Perhaps the
only parameters that get passed to the dispatcher via this method are the
unpacked values from the command?

```suggestion
func (d *Dispatcher) CancelMessage(directive string, message_id string, cancel_id string) error {
```

This forces all the unmarshalling and parameter validation into the client and
makes it very clear which message ID is to be cancelled.

Another option is to follow the pattern of the `Dispatch` method above, and expect
the entire `yggdrasil.Control` struct in the method:

```suggestion
func (d *Dispatcher) CancelMessage(control yggdrasil.Control) error {
```

This has the downside of unmarshalling the Command struct out of the control
message a second time (the first having occurred in the client when processing
the command name). I'm less inclined to take this approach, because this
introduces the potential to scope-creep the method into a larger "control
message handler" method. It's important to preserve it's responsibility as much
as possible.

So that being said, I think we should push all the message unpacking up to the
client and change the signature of this method as follows:

```suggestion
func (d *Dispatcher) Cancel(directive string, message_id string, cancel_id string) error {
```

Reply:
I wanted to avoid the CancelMessage to unmarshall again the control message. I'm OK to move this logic into the client. It seems a good idea.